File: scripts/check_daily_average.py
#!/usr/bin/env python3
"""
하루 평균 메일 확인 수 및 AI 검출 수를 확인하는 스크립트
"""
import os
import sys
from datetime import datetime, timedelta, timezone
from collections import defaultdict

# 프로젝트 루트를 경로에 추가
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from app.utils.state_store import FirestoreStateStore

def check_daily_average():
    """최근 30일간의 일별 평균 통계를 계산하고 출력"""
    try:
        store = FirestoreStateStore()
        db = store.db
        
        if not db:
            print("❌ Firestore 연결 실패")
            return
        
        # KST(UTC+9) 기준으로 날짜 계산
        kst = timezone(timedelta(hours=9))
        now = datetime.now(kst)
        
        # 최근 30일 데이터 조회
        days_to_analyze = 30
        days_ago_start = now - timedelta(days=days_to_analyze)
        days_ago_start_utc = days_ago_start.astimezone(timezone.utc)
        
        print(f"\n📊 최근 {days_to_analyze}일간 통계 분석 중...")
        print(f"기간: {days_ago_start.date()} ~ {now.date()} (KST)\n")
        
        # Firestore에서 이벤트 조회
        events_ref = db.collection("email_events")
        events_query = events_ref.where("timestamp", ">=", days_ago_start_utc)
        events_snapshot = events_query.stream()
        
        events_list = list(events_snapshot)
        print(f"총 이벤트 수: {len(events_list)}건\n")
        
        # 일별 집계
        daily_stats = defaultdict(lambda: {"total": 0, "ai_detected": 0})
        
        for doc in events_list:
            data = doc.to_dict()
            
            # 날짜 추출 (KST 기준)
            date_str = None
            if "created_at" in data and data["created_at"]:
                if hasattr(data["created_at"], "timestamp"):
                    event_date = datetime.fromtimestamp(data["created_at"].timestamp(), tz=kst)
                else:
                    event_date = data["created_at"].astimezone(kst)
                date_str = event_date.date().isoformat()
            elif "timestamp" in data and data["timestamp"]:
                if hasattr(data["timestamp"], "timestamp"):
                    event_date = datetime.fromtimestamp(data["timestamp"].timestamp(), tz=kst)
                else:
                    event_date = data["timestamp"].astimezone(kst)
                date_str = event_date.date().isoformat()
            
            if not date_str:
                continue
            
            # 일별 통계 업데이트
            daily_stats[date_str]["total"] += 1
            
            # AI 검출 여부 확인 (LLM 토큰이 있으면 AI 검출로 간주)
            has_ai_detection = (
                (data.get("llm_input_tokens") or 0) > 0 or
                (data.get("llm_output_tokens") or 0) > 0
            )
            if has_ai_detection:
                daily_stats[date_str]["ai_detected"] += 1
        
        # 평균 계산
        daily_stats_list = sorted(daily_stats.items())
        
        if not daily_stats_list:
            print("❌ 데이터가 없습니다.")
            return
        
        total_processed = sum(stats["total"] for _, stats in daily_stats_list)
        total_ai_detected = sum(stats["ai_detected"] for _, stats in daily_stats_list)
        
        avg_daily_processed = round(total_processed / len(daily_stats_list))
        avg_daily_ai_detected = round(total_ai_detected / len(daily_stats_list))
        
        # 결과 출력
        print("=" * 60)
        print("📈 하루 평균 통계")
        print("=" * 60)
        print(f"\n✅ 하루 평균 메일 확인 수: {avg_daily_processed:,}통")
        print(f"🤖 하루 평균 AI 검출 수: {avg_daily_ai_detected:,}통")
        print(f"\n📅 집계 기간: {len(daily_stats_list)}일")
        print(f"📊 총 처리 메일: {total_processed:,}통")
        print(f"🤖 총 AI 검출: {total_ai_detected:,}통")
        
        if daily_stats_list:
            print(f"\n📊 일별 상세 통계 (최근 10일):")
            print("-" * 60)
            for date_str, stats in daily_stats_list[-10:]:
                percentage = (stats["ai_detected"] / stats["total"] * 100) if stats["total"] > 0 else 0
                print(
                    f"{date_str}: 전체 {stats['total']:4d}통 | "
                    f"AI {stats['ai_detected']:4d}통 ({percentage:5.1f}%)"
                )
        
        print("\n" + "=" * 60 + "\n")
        
    except Exception as e:
        print(f"❌ 오류 발생: {e}")
        import traceback
        traceback.print_exc()
        sys.exit(1)

if __name__ == "__main__":
    check_daily_average()

File: admin-web/scripts/check-daily-average.ts
import * as admin from 'firebase-admin';
import * as dotenv from 'dotenv';
import * as path from 'path';

// 환경변수 로드
dotenv.config({ path: path.join(__dirname, '../.env.local') });

// Firebase Admin 초기화
if (!admin.apps.length) {
  const projectId = process.env.FIRESTORE_PROJECT_ID;
  const credentialsJson = process.env.GOOGLE_APPLICATION_CREDENTIALS_JSON_BASE64;

  if (!projectId) {
    console.error('FIRESTORE_PROJECT_ID 환경변수가 설정되지 않았습니다.');
    process.exit(1);
  }

  try {
    if (credentialsJson) {
      const decoded = Buffer.from(credentialsJson, 'base64').toString('utf-8');
      const serviceAccount = JSON.parse(decoded);
      admin.initializeApp({
        credential: admin.credential.cert(serviceAccount),
        projectId,
      });
    } else if (process.env.GOOGLE_APPLICATION_CREDENTIALS) {
      admin.initializeApp({
        credential: admin.credential.cert(process.env.GOOGLE_APPLICATION_CREDENTIALS),
        projectId,
      });
    } else {
      // 기본 인증 사용 (GCP 환경에서 실행 시)
      admin.initializeApp({
        projectId,
      });
    }
  } catch (error) {
    console.error('Firebase Admin initialization failed:', error);
    process.exit(1);
  }
}

const db = admin.firestore();

async function checkDailyAverage() {
  try {
    // KST(UTC+9) 기준으로 날짜 계산
    const now = new Date();
    const kstOffset = 9 * 60 * 60 * 1000;
    const kstNow = new Date(now.getTime() + kstOffset);
    
    // 최근 30일 데이터 조회
    const daysToAnalyze = 30;
    const kstDaysAgoStart = new Date(kstNow.getFullYear(), kstNow.getMonth(), kstNow.getDate() - daysToAnalyze);
    const daysAgoStart = new Date(kstDaysAgoStart.getTime() - kstOffset);
    
    console.log(`\n📊 최근 ${daysToAnalyze}일간 통계 분석 중...`);
    console.log(`기간: ${kstDaysAgoStart.toISOString().split('T')[0]} ~ ${kstNow.toISOString().split('T')[0]} (KST)\n`);
    
    const recentEventsSnapshot = await db.collection("email_events")
      .where("timestamp", ">=", daysAgoStart)
      .get();
    
    console.log(`총 이벤트 수: ${recentEventsSnapshot.size}건\n`);
    
    // 일별 집계
    const dailyStatsMap = new Map<string, { total: number; aiDetected: number }>();
    
    recentEventsSnapshot.docs.forEach(doc => {
      const data = doc.data();
      
      // 날짜 추출 (KST 기준)
      let dateStr: string;
      if (data.created_at && data.created_at.toDate) {
        const eventDate = new Date(data.created_at.toDate().getTime() + kstOffset);
        dateStr = eventDate.toISOString().split('T')[0];
      } else if (data.timestamp && data.timestamp.toDate) {
        const eventDate = new Date(data.timestamp.toDate().getTime() + kstOffset);
        dateStr = eventDate.toISOString().split('T')[0];
      } else {
        return; // 날짜 정보 없으면 스킵
      }
      
      // 일별 통계 초기화
      if (!dailyStatsMap.has(dateStr)) {
        dailyStatsMap.set(dateStr, { total: 0, aiDetected: 0 });
      }
      
      const dailyStats = dailyStatsMap.get(dateStr)!;
      dailyStats.total++;
      
      // AI 검출 여부 확인 (LLM 토큰이 있으면 AI 검출로 간주)
      const hasAiDetection = (data.llm_input_tokens || 0) > 0 || (data.llm_output_tokens || 0) > 0;
      if (hasAiDetection) {
        dailyStats.aiDetected++;
      }
    });
    
    // 평균 계산
    const dailyStatsArray = Array.from(dailyStatsMap.entries())
      .map(([date, stats]) => ({ date, ...stats }))
      .sort((a, b) => a.date.localeCompare(b.date));
    
    const avgDailyProcessed = dailyStatsArray.length > 0 
      ? Math.round(dailyStatsArray.reduce((sum, stats) => sum + stats.total, 0) / dailyStatsArray.length)
      : 0;
    const avgDailyAiDetected = dailyStatsArray.length > 0
      ? Math.round(dailyStatsArray.reduce((sum, stats) => sum + stats.aiDetected, 0) / dailyStatsArray.length)
      : 0;
    
    // 결과 출력
    console.log('='.repeat(60));
    console.log('📈 하루 평균 통계');
    console.log('='.repeat(60));
    console.log(`\n✅ 하루 평균 메일 확인 수: ${avgDailyProcessed.toLocaleString()}통`);
    console.log(`🤖 하루 평균 AI 검출 수: ${avgDailyAiDetected.toLocaleString()}통`);
    console.log(`\n📅 집계 기간: ${dailyStatsArray.length}일`);
    console.log(`📊 총 처리 메일: ${recentEventsSnapshot.size.toLocaleString()}통`);
    console.log(`🤖 총 AI 검출: ${dailyStatsArray.reduce((sum, stats) => sum + stats.aiDetected, 0).toLocaleString()}통`);
    
    if (dailyStatsArray.length > 0) {
      console.log(`\n📊 일별 상세 통계 (최근 10일):`);
      console.log('-'.repeat(60));
      dailyStatsArray.slice(-10).forEach(stat => {
        const percentage = stat.total > 0 ? ((stat.aiDetected / stat.total) * 100).toFixed(1) : '0.0';
        console.log(`${stat.date}: 전체 ${stat.total.toString().padStart(4)}통 | AI ${stat.aiDetected.toString().padStart(4)}통 (${percentage}%)`);
      });
    }
    
    console.log('\n' + '='.repeat(60) + '\n');
    
  } catch (error) {
    console.error('❌ 오류 발생:', error);
    process.exit(1);
  }
}

checkDailyAverage().then(() => {
  process.exit(0);
}).catch((error) => {
  console.error('❌ 실행 오류:', error);
  process.exit(1);
});
